File: harness/go-0.12/internal/reason/reason.go
// Package reason reduces valiss-go's descriptive errors to the stable
// spec/SPEC-1.md §7 reason codes the interop contract asserts on rejection.
// It mirrors the substring table of valiss-go's conformance runner
// (conformance/conformance_test.go), extended with the Verifier-stage codes
// (§7.2 windows/epoch and §7.3 credential checks) that the request-level
// harness exercises and the vector-level runner does not.
package reason

import "strings"

// mapping is one §7 code and the stable fragments of the valiss-go error
// strings that reduce to it. Fragments are chosen to be mutually unambiguous
// across the table (see the conformance runner for the two near-collisions:
// "token signature:" vs "token signature verification failed", and the
// message-chain wrong-issuer string, which the more specific §7.4 code wins).
type mapping struct {
	code       string
	substrings []string
}

// table is ordered with the most specific request-stage fragments first;
// every fragment matches exactly one code, so the order is defensive, not
// load-bearing. References are to valiss-go at v0.12.0.
var table = []mapping{
	// §7.3 request / credential
	{"replay", []string{"nonce already seen"}},                                   // verifier.go:379
	{"nonce_required", []string{"request nonce required"}},                       // verifier.go:376
	{"bad_request_signature", []string{"request signature verification failed"}}, // sign.go:79
	{"bad_signature_encoding", []string{"bad request signature encoding"}},       // sign.go:71
	{"skew", []string{"skew window", "bad request timestamp"}},                   // sign.go:64-69
	{"not_bearer", []string{"request signature required"}},                       // verifier.go:367
	{"not_allowlisted", []string{"account token not recognized"}},                // verifier.go:339
	{"unknown_operator", []string{"no trusted operator"}},                        // verifier.go:310; message.go:273
	{"no_resolver", []string{"no account token resolver"}},                       // verifier.go:283
	{"operator_misconfigured", []string{"operator token misconfigured"}},         // verifier.go:276
	{"missing", []string{"missing credentials", "no token markers"}},             // verifier.go:280; creds.go:99
	{"extension_invalid", []string{"decode extension"}},                          // verifier.go:385-392

	// §7.2 token semantics
	{"epoch_mismatch", []string{"token epoch"}},  // verifier.go:329/348; message.go:334-342
	{"expired", []string{"token expired"}},       // verifier.go:323/333/351; message.go windows
	{"not_yet_valid", []string{"not yet valid"}}, // verifier.go:326/336/354
	{"wrong_type", []string{
		"not an operator token", "not an account token",
		"not a user token", "not a message token",
	}}, // token.go:356/381/406; message.go:294
	{"chain_user_mismatch", []string{"not signed by the chain's user key"}}, // message.go:324 (wins over wrong_issuer)
	{"wrong_issuer", []string{
		"not self-signed by the expected",    // operator (token.go:359)
		"not signed by the expected issuer",  // account (token.go:384)
		"not signed by the expected account", // user (token.go:409)
		"not self-signed by its user key",    // message iss != sub (message.go:297)
	}},
	{"wrong_subject_role", []string{"subject is not"}}, // token.go:362/387/412; message.go:300

	// §7.4 message-specific
	{"no_chain", []string{"carries no chain"}},                      // message.go:22/305
	{"chain_mismatch", []string{"differs from the supplied chain"}}, // message.go:309
	{"wrong_audience", []string{"message token audience"}},          // message.go:372
	{"checksum_missing", []string{"carries no checksum"}},           // message.go:377/383
	{"checksum_mismatch", []string{"payload checksum mismatch"}},    // message.go:379

	// §7.1 envelope / decode
	{"bad_signature", []string{"token signature verification failed"}},                   // token.go:251
	{"unsupported_type", []string{"unsupported token type"}},                             // token.go:201
	{"unsupported_version", []string{"unsupported wire version", "unsupported version"}}, // token.go:219; creds.go:135
	{"bad_issuer_key", []string{"token issuer:"}},                                        // token.go:243
	{"malformed", []string{
		"malformed token",  // not three parts (token.go:186)
		"token header:",    // header not base64url or not JSON (token.go:190-199)
		"token claims:",    // payload not base64url or not JSON (token.go:229-241)
		"token signature:", // signature not base64url (token.go:247-249)
		// creds envelope (creds.go between/checkVersion)
		"not closed", "unexpected content", "no content before",
	}},
}

// Code reduces a valiss-go verification error to its §7 reason code. An
// error outside the table falls back to "malformed": §7 requires every
// failure to reduce to some code, and an unmapped string is a harness bug
// best surfaced by a matrix mismatch, not a crash.
func Code(err error) string {
	msg := err.Error()
	for _, m := range table {
		for _, s := range m.substrings {
			if strings.Contains(msg, s) {
				return m.code
			}
		}
	}
	return "malformed"
}

File: harness/py-0.8-asgi/interopharness/reason.py
"""Reduce the shipped ASGI middlewares' rejection text to the stable
spec/SPEC-1.md §7 reason codes the interop contract asserts on rejection.

valiss-py carries first-class reason codes (``ValissError.reason``), but the
ASGI middlewares render rejections as ``str(exc)`` — the human-readable
message the library documents as illustrative — and expose neither the code
nor a rendering hook to the wrapping application. So this entry re-derives
the code from the message the way the Go framework-adapter entries do against
raw Go errors (their ``internal/reason`` substring tables). The fragments
below are the valiss-py texts frozen by this entry's exact ``valiss==0.8.0``
pin, plus the fixed strings the middlewares render themselves (the
missing-token and chain-required rejections).
"""

from __future__ import annotations

# The table pairs one §7 code with the stable fragments of the valiss-py
# error strings that reduce to it. It is ordered with the most specific
# request-stage fragments first; every fragment matches exactly one code, so
# the order is defensive, not load-bearing. References are to valiss-py at
# v0.8.0.
_TABLE: list[tuple[str, list[str]]] = [
    # §7.3 request / credential
    ("replay", ["nonce already seen"]),                                   # verifier.py
    ("nonce_required", ["request nonce required"]),                       # verifier.py
    ("bad_request_signature", [
        "request signature verification failed",                          # token.py verify_signature
        "bad subject public key",                                         # token.py verify_signature
    ]),
    ("bad_signature_encoding", ["bad request signature encoding"]),       # token.py verify_signature
    ("skew", ["skew window", "bad request timestamp"]),                   # token.py verify_signature
    ("not_bearer", ["request signature required"]),                       # verifier.py
    ("not_allowlisted", ["account token not recognized"]),                # verifier.py
    ("unknown_operator", ["no trusted operator"]),                        # verifier.py; message.py
    ("no_resolver", ["no account token resolver"]),                       # verifier.py
    ("operator_misconfigured", ["operator token misconfigured"]),         # verifier.py
    ("validator_rejected", ["validator rejected"]),                       # verifier.py
    ("missing", [
        "missing credentials",                                            # httpauth._server / verifier.py
        "missing message token",                                          # httpsig.asgi middleware
    ]),
    ("extension_invalid", [
        "decode extension",                                               # token.py ext_of
        "no http extension",                                              # httpauth.extension authorize_ext
        "does not permit",                                                # httpauth.extension authorize_ext
    ]),

    # §7.2 token semantics
    ("epoch_mismatch", ["token epoch"]),                                  # verifier.py; message.py
    ("expired", ["token expired"]),                                       # verifier.py; message.py windows
    ("not_yet_valid", ["not yet valid"]),                                 # verifier.py; message.py windows
    ("wrong_type", [
        "not an operator token", "not an account token",
        "not a user token", "not a message token",
    ]),                                                                   # token.py; message.py
    ("chain_user_mismatch", ["not signed by the chain's user key"]),      # message.py (wins over wrong_issuer)
    ("wrong_issuer", [
        "not self-signed by the expected",                                # operator (token.py)
        "not signed by the expected issuer",                              # account (token.py)
        "not signed by the expected account",                             # user (token.py)
        "not self-signed by its user key",                                # message iss != sub (message.py)
    ]),
    ("wrong_subject_role", ["subject is not"]),                           # token.py; message.py

    # §7.4 message-specific. The httpsig middleware answers a chainless
    # token with its own fixed "message token chain required" text instead
    # of the library's "carries no chain" error, so both fragments reduce
    # to no_chain.
    ("no_chain", ["carries no chain", "message token chain required"]),   # message.py; _msgtransport.py
    ("chain_mismatch", ["differs from the supplied chain"]),              # message.py
    ("wrong_audience", ["message token audience"]),                       # message.py
    ("checksum_missing", ["carries no checksum"]),                        # message.py
    ("checksum_mismatch", ["payload checksum mismatch"]),                 # message.py

    # §7.1 envelope / decode
    ("bad_signature", ["token signature verification failed"]),           # token.py _decode_v1
    ("unsupported_type", ["unsupported token type"]),                     # token.py _peek_version
    ("unsupported_version", ["unsupported wire version"]),                # token.py _decode_token
    ("bad_issuer_key", ["token issuer:"]),                                # token.py _decode_v1
    ("malformed", [
        "malformed token",                                                # not three parts / header version
        "token header:",                                                  # header not base64url / not JSON
        "token claims:",                                                  # payload not base64url / wrong shapes
    ]),
]


def code(message: str) -> str:
    """Reduce a middleware rejection's text to its §7 reason code. Text
    outside the table falls back to "malformed": §7 requires every failure to
    reduce to some code, and an unmapped string is a harness bug best
    surfaced by a matrix mismatch, not a crash."""
    for reason_code, fragments in _TABLE:
        for fragment in fragments:
            if fragment in message:
                return reason_code
    return "malformed"
